File: Objc Example/Objc Example/ViewController.m
//
//  Created by Suppy.io
//

#import "ViewController.h"
#import "SubtitleTableViewCell.h"
#import "ConfigKey.h"

static NSString *kCellIdentifier = @"cellIdentifier";

@interface ViewController () 
@end

@implementation ViewController {
    UITableView             *tableView;
    NSMutableDictionary     *data;
    NSArray                 *configKeys;
}

- (void)refreshData
{
    dispatch_async(dispatch_get_main_queue(), ^(void){
        [self loadFromUserDefaults];
        [self->tableView reloadData];
    });
}

- (void)loadFromUserDefaults
{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];

    [data removeAllObjects];

    for(NSString *key in configKeys)
    {
        if([key isEqualToString:kPrivacyPolicy]) {
            NSURL *url = [defaults URLForKey:key];
            [data setValue:url.absoluteString forKey:key];
        } else if([key isEqualToString:kProductList]) {
            [data setValue:[[defaults arrayForKey:key] componentsJoinedByString:@", "] forKey:key];
        } else if([key isEqualToString:kFeatureXEnabled]) {
            BOOL isEnabled = [defaults boolForKey: key];
            NSString *boolStr = isEnabled ? @"True" : @"False";
            [data setValue:boolStr forKey:key];
        } else {
            [data setValue:[defaults stringForKey:key] forKey:key];
        }
    }
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    data = [NSMutableDictionary dictionary];
    configKeys = [ConfigKey singleton].globalArray;

    tableView = [[UITableView alloc] init];

    [self.view addSubview:tableView];

    tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [[tableView topAnchor] constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [[tableView bottomAnchor] constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
    [[tableView leftAnchor] constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [[tableView rightAnchor] constraintEqualToAnchor:self.view.rightAnchor].active = YES;

    tableView.delegate = self;
    tableView.dataSource = self;

    self.view.backgroundColor = UIColor.whiteColor;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [ConfigKey singleton].globalArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellIdentifier];

    if(cell == nil) {
        cell = [[SubtitleTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:kCellIdentifier];
    }

    NSString *key = configKeys[indexPath.row];
    cell.textLabel.text = [data valueForKey:key];
    cell.detailTextLabel.text = key;

    if([key isEqualToString:kBackgroundColor]) {
        NSString *color =  [data valueForKey:key];
        if([color isEqualToString:@"white"]) {
            cell.backgroundColor = UIColor.whiteColor;
        } else if([color isEqualToString:@"red"]) {
            cell.backgroundColor = UIColor.redColor;
        } else if([color isEqualToString:@"blue"]) {
            cell.backgroundColor = UIColor.blueColor;
        } else if([color isEqualToString:@"green"]) {
            cell.backgroundColor = UIColor.greenColor;
        } else if([color isEqualToString:@"yellow"]) {
            cell.backgroundColor = UIColor.yellowColor;
        } else {
            cell.backgroundColor = UIColor.whiteColor;
        }
    }

    return cell;
}

@end

File: Swift Example/Swift Example/ViewController.swift
//
//  Created by Suppy.io
//

import UIKit

private let cellIdentifier = "cell"

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    private let tableView = UITableView()
    private var data = [(key: String, value: String?)]()

    func refreshData() {
        DispatchQueue.main.async { [self] in
            loadFromUserDefaults()
            tableView.reloadData()
        }
    }

    func loadFromUserDefaults() {
        let defaults = UserDefaults.standard

        data.removeAll()

        ConfigKey.allCases.forEach { (key) in
            switch key {
            case .privacyPolicy:
                let url = defaults.url(forKey: key.rawValue)
                data.append((key: key.rawValue, value: url?.absoluteString))
            case .productList:
                let array = defaults.array(forKey: key.rawValue) as? [String]
                data.append((key: key.rawValue, value: array?.joined(separator: ", ")))
            default:
                data.append((key: key.rawValue, value: defaults.string(forKey: key.rawValue)))
            }
        }

        data.sort { $0.key < $1.key }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.view.addSubview(tableView)

        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.topAnchor.constraint(equalTo: self.view.topAnchor).isActive = true
        tableView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor).isActive = true
        tableView.leftAnchor.constraint(equalTo: self.view.leftAnchor).isActive = true
        tableView.rightAnchor.constraint(equalTo: self.view.rightAnchor).isActive = true

        tableView.dataSource = self
        tableView.delegate = self

        tableView.register(SubtitleTableViewCell.self, forCellReuseIdentifier: cellIdentifier)
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return data.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)

        let row = data[indexPath.row]
        let key = row.key
        let value = row.value

        cell.textLabel?.text = value
        cell.detailTextLabel?.text = key

        if key == ConfigKey.backgroundColor.rawValue,
           let value = value
        {
            switch value {
            case "white": cell.backgroundColor = UIColor.white
            case "red": cell.backgroundColor = UIColor.red
            case "blue": cell.backgroundColor = UIColor.blue
            case "green": cell.backgroundColor = UIColor.green
            case "yellow": cell.backgroundColor = UIColor.yellow
            default:
                cell.backgroundColor = UIColor.white
            }
        }
        return cell
    }
}

